ee/agent/startup/startup.go

```diff
@@ -0,0 +1,94 @@
+// Package startup provides access to and manages storage of startup data:
+// flags/config values/settings that launcher needs during initialization,
+// before the knapsack is available.
+package startup
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+
+	_ "github.com/golang-migrate/migrate/v4/database/sqlite"
+	"github.com/kolide/launcher/ee/agent/flags/keys"
+	agentsqlite "github.com/kolide/launcher/ee/agent/storage/sqlite"
+	"github.com/kolide/launcher/ee/agent/types"
+	_ "modernc.org/sqlite"
+)
+
+// GetStartupValue retrieves the value for the given flagKey from the startup database
+// located in the given rootDirectory. It wraps creation and closing of the sqlite store.
+func GetStartupValue(ctx context.Context, rootDirectory string, flagKey string) (string, error) {
+	store, err := agentsqlite.NewStore(ctx, rootDirectory)
+	if err != nil {
+		return "", fmt.Errorf("opening startup db in %s: %w", rootDirectory, err)
+	}
+	defer store.Close()
+
+	flagValue, err := store.Get([]byte(flagKey))
+	if err != nil {
+		return "", fmt.Errorf("getting flag value %s: %w", flagKey, err)
+	}
+
+	return string(flagValue), nil
+}
+
+// startupDatabase records agent flags and their current values,
+// responding to updates as a types.FlagsChangeObserver
+type startupDatabase struct {
+	kvStore     *agentsqlite.SqliteStore
+	knapsack    types.Knapsack
+	storedFlags map[keys.FlagKey]func() string // maps the agent flags to their knapsack getter functions
+}
+
+// NewStartupDatabase returns a new startup database, creating and initializing
+// the database if necessary.
+func NewStartupDatabase(ctx context.Context, knapsack types.Knapsack) (*startupDatabase, error) {
+	store, err := agentsqlite.NewStore(ctx, knapsack.RootDirectory())
+	if err != nil {
+		return nil, fmt.Errorf("opening startup db in %s: %w", knapsack.RootDirectory(), err)
+	}
+
+	s := &startupDatabase{
+		kvStore:  store,
+		knapsack: knapsack,
+		storedFlags: map[keys.FlagKey]func() string{
+			keys.UpdateChannel: func() string { return knapsack.UpdateChannel() },
+		},
+	}
+
+	// Attempt to ensure flags are up-to-date
+	if err := s.setFlags(ctx); err != nil {
+		s.knapsack.Slogger().Log(ctx, slog.LevelWarn, "could not set flags", "err", err)
```

Comment:
just a nit, but when ever we add a kvp beyond msg, I prefer it to look like this ... but that is just my preference, can live with either way

```suggestion
		s.knapsack.Slogger().Log(ctx, slog.LevelWarn, 
		    "could not set flags", 
		    "err", err,
		  )
```

Reply:
I prefer that as well! I'll update.